KAFKA-2366: Copycat

ewencp opened the pull request:
This is an initial patch implementing the basics of Copycat for KIP-26.

The intent here is to start a review of the key pieces of the core API and get a reasonably functional, baseline, non-distributed implementation of Copycat in place to get things rolling. The current patch has a number of known issues that need to be addressed before a final version:
- Some build-related issues. Specifically, requires some locally-installed dependencies (see below), ignores checkstyle for the runtime data library because it's lifted from Avro currently and likely won't last in its current form, and some Gradle task dependencies aren't quite right because I haven't gotten rid of the dependency on `core` (which should now be an easy patch since new consumer groups are in a much better state).
- This patch currently depends on some Confluent trunk code because I prototyped with our Avro serializers w/ schema-registry support. We need to figure out what we want to provide as an example built-in set of serializers. Unlike core Kafka where we could ignore the issue, providing only ByteArray or String serializers, this is pretty central to how Copycat works.
- This patch uses a hacked up version of Avro as its runtime data format. Not sure if we want to go through the entire API discussion just to get some basic code committed, so I filed KAFKA-2367 to handle that separately. The core connector APIs and the runtime data APIs are entirely orthogonal.
- This patch needs some updates to get aligned with recent new consumer changes (specifically, I'm aware of the ConcurrentModificationException issue on exit). More generally, the new consumer is in flux but Copycat depends on it, so there are likely to be some negative interactions.
- The layout feels a bit awkward to me right now because I ported it from a Maven layout. We don't have nearly the same level of granularity in Kafka currently (core and clients, plus the mostly ignored examples, log4j-appender, and a couple of contribs). We might want to reorganize, although keeping data+api separate from runtime and connector plugins is useful for minimizing dependencies.
- There are a variety of other things (e.g., I'm not happy with the exception hierarchy/how they are currently handled, TopicPartition doesn't really need to be duplicated unless we want Copycat entirely isolated from the Kafka APIs, etc), but I expect those we'll cover in the review.

Before commenting on the patch, it's probably worth reviewing https://issues.apache.org/jira/browse/KAFKA-2365 and https://issues.apache.org/jira/browse/KAFKA-2366 to get an idea of what I had in mind for a) what we ultimately want with all the Copycat patches and b) what we aim to cover in this initial patch. My hope is that we can use a WIP patch (after the current obvious deficiencies are addressed) while recognizing that we want to make iterative progress with a bunch of subsequent PRs.

ewencp committed:
Add copycat-data and copycat-api

ewencp committed:
Add copycat-avro and copycat-runtime

ewencp committed:
Add Copycat file connector.

ewencp committed:
Add CLI tools for Copycat.

ewencp committed:
Add some helpful Copycat-specific build and test targets that cover all Copycat packages.

ijuma commented:
Great to see this!

I did a partial review of the first commit. I didn't continue because I am a bit unsure about which parts are meant to be discussed separately as part of KAFKA-2367 and also the level of review we want to do at this stage (e.g. do we care about code convention issues).

I understand the desire to get some initial code merged and to iterate in subsequent PRs. At the same time, it's a significant chunk of code. Once it's merged, the focus will shift and it's unclear to me exactly which parts will be rewritten and which parts are here to stay.

Thoughts?

ewencp commented:
Yeah, hard to say how we should balance just getting some code in place vs. actually being happy with the state of it all. At a minimum, definitely point out code convention stuff since those are trivial to fix anyway. This code was originally under Confluent's code styling and I reformatted, so there are likely a bunch of places I missed cleaning up. One thing I've noticed consistently is that since we use shorter lines, formatting often ends up a bit messier and spread across multiple lines and automatic reformatting won't fix that.

For KAFKA-2367, the intent in that JIRA was to focus on the API currently in copycat-data, which is curruently just a stripped down Avro with all the serialization code removed. I had initially left out "data" from the title of the JIRA which made it sound a bit different, despite the description being clearer. It'd be a good idea to come to some agreement on the APIs in copycat-api in this JIRA since they affect many of the other  JIRAs (and the runtime model for connectors).

ewencp committed:
Switch to using new consumer coordinator instead of manually assigning partitions. Remove dependency of copycat-runtime on core.

ewencp committed:
Remove duplicated TopicPartition implementation.

ewencp committed:
Remove offset serializers, instead reusing the existing serializers and removing schema projection support.

ewencp committed:
Merge remote-tracking branch 'origin/trunk' into copycat

ewencp committed:
Replace Avro serializer with JSON serializer.

ewencp committed:
Add log4j settings for Copycat.

ewencp committed:
Fix sink task offset commit concurrency issue by moving it to the worker thread and waking up the consumer to ensure it exits promptly.

ewencp commented:
I've addressed the major issues with the original patch now:
- Replaced the Avro serializer with a JSON one. This requires wrapping the actual payload in an envelope so we can ship the schema with it as well. I think we can make this optional (ideally in a later patch) by allowing the data API to express a catch-all schema, include a flag for serializers to allow them to read "raw" JSON and attach the catch-all schema to it. This will allow using existing JSON data, avoids the overhead , and will only be incompatible with copycat plugins that really require the schema (e.g. the HDFS connector may not be able to write that data in all formats).
- Fixed the threading issue with the new consumer.
- Simplified the serializers by getting rid of the specialized offset serializers and just reuse the existing producer/consumer serializer interfaces.
- Switched to using consumer coordinator. This fixed a dependency on core.

This also gets it into a state where we don't have an external dependency. It should be easily testable for everyone now.

ewencp committed:
Address review comments, clean up some code styling.

ewencp committed:
Remove most of the Avro-based mock runtime data API, only preserving enough schema functionality to support basic primitive types for an initial patch.

ewencp commented:
https://issues.apache.org/jira/browse/KAFKA-2367 is intended to address the data API so this patch doesn't get held up while we agree on what that API should look like. But it was pointed out to me that such a significant chunk of this patch's code was in copycat-data that it distracts from the rest of the patch, which should be the focus.

To that end, I've removed the majority of the API. What I preserved still isn't intended to be a final API, but we need some basic implementation of a `Schema` API to support a basic `Serializer` implementation to at least make a basic version using primitive types functionally. Hopefully this can redirect attention to the new Copycat APIs that are introduced in this patch.

Still, it might be worth pointing out that the `Schema` implementation is nearly 1/3 of this entire patch still -- we still have a significant chunk of code here, but the patch is quite a bit smaller than it looks...

ewencp committed:
Style cleanup

ewencp committed:
Minor cleanup

ewencp committed:
Remove Copycat reflection utils, use existing Utils and ConfigDef functionality from clients package.

ewencp committed:
Normalize naming to use partition for both source and Kafka, adjusting naming in CopycatRecord classes to clearly differentiate.

ewencp committed:
Split Copycat converter option into two options for key and value.

ewencp committed:
Make Converter generic to match serializers since some serialization formats do not require a base class of Object; update many other classes to have generic key and value class type parameters to match this change.

ewencp commented:
@ijuma (cc @nehanarkhede since we discussed these settings recently) The question about generics was a good one. I've updated `Converter` and added generics as necessary based on the fall out. However, I noticed two things.

First, with the current code this ends up bubbling all the way up to the `Worker` class. This makes sense since it syncs with the global converter/serializer/deserializer configs. However, now I'm thinking that might be a problem. Many connectors may want to use complex keys, but users may have existing data with simple keys, e.g. strings (not wrapped in Avro/Protobufs/Thrift/JSON/whatever). Currently we can't support both of these running in the same process since they require different deserializers/converters and the serializer/deserializer/converter are all set globally. This seems like it could be a problem for clustered mode where you want a lot of potentially diverse connectors to be running under the same workers.

Second, there's a similar issue with offsets. We got rid of `offset.key.serializer` and `offset.value.serializer` because there were a lot of settings. But if you want to be able to decode plain string keys but also decode complex offsets (as source connectors will need to do) you'll end up with a similar conflict.

We could definitely add the necessary settings and restructure the code a bit to support this, but it seems like it's going to add a lot of complexity. We'd need to at least add the offset key/value serializer/deserializer settings globally _and_ add per-connector settings for serializer/deserializer. The latter seems like it has the wost impact since it means every connector has to specify settings like this:

```
key.converter=org.apache.kafka.copycat.json.JsonConverter
value.converter=org.apache.kafka.copycat.json.JsonConverter
key.serializer=org.apache.kafka.copycat.json.JsonSerializer
value.serializer=org.apache.kafka.copycat.json.JsonSerializer
key.deserializer=org.apache.kafka.copycat.json.JsonDeserializer
value.deserializer=org.apache.kafka.copycat.json.JsonDeserializer
```

With the new producer we also moved to a policy of not providing defaults for things like this because they lead to unclear error messages when something goes wrong such that users can't figure out how to fix the problem. We could simplify things by providing a worker/cluster-level default, but that seems like it could have the same problem as providing defaults in the producer.

gwenshap commented on build.gradle:
```diff
@@ -496,3 +506,238 @@ project(':log4j-appender') {
+    archives.extendsFrom (testCompile)
+  }
+
+  /* FIXME
```
Can you add a JIRA # where you will fix this? It helps with tracking.

gwenshap commented on clients/checkstyle/import-control.xml:
```diff
@@ -0,0 +1,102 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+--> 
+<import-control pkg="org.apache.kafka">
```
Why are we re-adding this file?

ewencp replied:
Because it was an accident. My IntelliJ apparently has some issues and requires extra copies of these files... Removed in follow up commit

gwenshap commented on clients/src/main/java/org/apache/kafka/common/utils/Utils.java:
```diff
@@ -271,7 +271,7 @@ public static void sleep(long ms) {
      */
-    public static Object newInstance(Class<?> c) {
+    public static <T> T newInstance(Class<T> c) {
         try {
```
Not your code, but this it a public method that will go up in flame if c == null

gwenshap commented on config/copycat-console-sink.properties:
```diff
@@ -0,0 +1,19 @@
+# limitations under the License.
+
+name=local-console-sink
+connector.class=org.apache.kafka.copycat.file.FileStreamSinkConnector
```
Can be a followup, but Flume allows "aliasing" sources/sinks and using a friendly name instead of full class name. Worth copying IMO.

gwenshap commented on copycat-api/src/main/java/org/apache/kafka/copycat/connector/Connector.java:
```diff
@@ -0,0 +1,121 @@
+     * This version is only used to recover from failures.
+     * </p>
+     * <p>
+     * The default implementation ignores the provided simply ignores the provided Task
```
Nit: this sentence was slightly damaged in editting

gwenshap commented on copycat-api/src/main/java/org/apache/kafka/copycat/connector/Connector.java:
```diff
@@ -0,0 +1,121 @@
+     * @param props configuration settings
+     * @throws CopycatException
+     */
+    public abstract void start(Properties props) throws CopycatException;
```
In Flume we separate start method (which is not retried) from configure (which we retry until success). Things like "connect to database" which can fail temporarily, go in "configure".

Maybe this will be useful here?

ewencp replied:
Hmm, maybe related to this is that it may not be clear right now where a connector developer should put code that may potentially block. Right now the way you'd do that is allocate your own thread and run something in the background (e.g. if you need to monitor for changes to the tables in a source database). Maybe we should formalize that so there is `configure(props)` and `run()` and Copycat handles allocating that thread.

However, not all connectors will need a permanently running thread (and may not need to do any checks after initial configuration), so I'm not sure requiring it for all connectors is a good idea. And I'm wondering if we should provide utilities for doing things like periodically polling.

gwenshap commented on copycat-api/src/main/java/org/apache/kafka/copycat/connector/CopycatRecord.java:
```diff
@@ -0,0 +1,100 @@
+ * notion of offset, it is not included here because they differ in type.
+ * </p>
+ */
+public abstract class CopycatRecord {
```
Why do we need a separate class then?

ewencp replied:
This is just inheritance for code reuse -- avoiding duplicating all this boilerplate. Happy to get rid of this class if reusing code this way is not preferred.

gwenshap replied:
All good, I got confused while reading.

gwenshap commented on copycat-api/src/main/java/org/apache/kafka/copycat/sink/SinkTaskContext.java:
```diff
@@ -0,0 +1,42 @@
+public abstract class SinkTaskContext {
+    private Map<TopicPartition, Long> offsets;
+
+    public SinkTaskContext() {
```
Please document the methods and when / why would the connector developer use them

gwenshap commented on copycat-api/src/main/java/org/apache/kafka/copycat/source/SourceConnector.java:
```diff
@@ -0,0 +1,27 @@
+ * SourceConnectors implement the connector interface to pull data from another system and send
+ * it to Kafka.
+ */
+public abstract class SourceConnector extends Connector {
```
This doesn't require TOPIC_CONFIG?

ewencp replied:
Interesting question. Some sources may need it, but others might auto-generate topic names. For example, a JDBC connector may just construct topic names as something like `<db_name>-<table_name>`. So I'm not sure we want to _require_ a TOPIC_CONFIG here.

However, on a related note, we may need to have some interface for getting the topics being published to. This is important if you would want to do something like combine an embedded copycat job with a kstream job -- to run the kstream job you need the set of topics that are being written to, which, as described above, might just be some dynamic set that can't be listed until the connector is running.

gwenshap replied:
Good point. Where do you see this interface living?

ewencp replied:
I'm not quite sure. One option would be to require SourceConnectors/SourceTasks to implement a method to get this information. However, users of the embedded API wouldn't be creating those classes themselves, just submitting a job, so they couldn't call this directly. I think we'd want chain it through other APIs all the way up to whatever top-level class is instantiated for embedded mode (i.e. the Copycat class that you would call run() and addConnector() on).

gwenshap commented on copycat-api/src/main/java/org/apache/kafka/copycat/source/SourceRecord.java:
```diff
@@ -0,0 +1,101 @@
+        return sourcePartition;
+    }
+
+    public Object getSourceOffset() {
```
I imagined all SourceConnectors will need to persist the source offset periodically (checkpoint or commit).
Do we need to expose a method for this?

ewencp replied:
They do need to commit periodically, although so far I haven't found a need to actually put this under the connectors control at all. The entire process can be managed entirely by the framework, which just respects whatever interval the connector config requests. However, this does mean every single message sent must also have a source offset filled in (assuming you want to track offsets, something like metrics might just leave it null).

I thought about cases like transactions where maybe it would be useful for the connector to indicate that it is at a good place to flush and commit, but I'm not sure it actually buys you anything since you still have to deal with half-finished transactions in case a process crashes. The only scenario I can think of where it is absolutely required is if you cannot come up with unique offsets for every message, but I haven't come up with a real use case where that is a problem.

gwenshap replied:
Thought about it again, and I agree with you - periodic commits should be good enough.

gwenshap commented on copycat-api/src/main/java/org/apache/kafka/copycat/source/SourceTaskContext.java:
```diff
@@ -0,0 +1,38 @@
+    /**
+     * Get the OffsetStorageReader for this SourceTask.
+     */
+    public OffsetStorageReader getOffsetStorageReader() {
```
Ah, I think I found it ;)

gwenshap commented on copycat-api/src/main/java/org/apache/kafka/copycat/storage/Converter.java:
```diff
@@ -0,0 +1,42 @@
+/**
+ * The Converter interface provides support for translating between Copycat's runtime data format
+ * and the "native" runtime format used by the serialization layer. This is used to translate
+ * two types of data: records and offsets. The (de)serialization is performed by a separate
```
We mean source offsets here? because sink offsets are serialized the way Kafka Offset manager needs them, right?

ewencp replied:
Yes, this really only applies to source offsets currently.

One thing that has bothered me is the inconsistency between how offsets are handled for source and sink tasks. I'm also wondering if there are any cases where storing additional data in a sink task offset commit would be useful. We could use the metadata field once that's exposed via the API for that, but we'd also have to make it available to Copycat somehow. In particular, I'm wondering if there is ever a case where you can get better delivery semantics to the destination system if you, e.g., get some ID back each time you send a message to it, and if you record that ID you might be able to get the destination system to do dedup -- however, I haven't come up with a system where this actually works yet :)

The other drawback to having them managed in different ways is that tooling needs to be different for source tasks vs. sink tasks. e.g., anything like burrow that tries to process the offsets topics will need to have two different implementations.

gwenshap replied:
HBase has the CheckAndPut API that seem like could do dedup if we provide the correct thing to "check".

gwenshap commented on copycat-api/src/main/java/org/apache/kafka/copycat/connector/Connector.java:
```diff
@@ -0,0 +1,121 @@
+     * @param props new configuration settings
+     * @throws CopycatException
+     */
+    public void reconfigure(Properties props) throws CopycatException {
```
Configuration of a connector will be "small data". We could force connectors to expose:
1. Version ID
2. update(from version) that will read configuration in known version and convert it to current configurations, storing a newer version.

If the current connector plugin version doesn't match stored configuration, we can <do something>. 

What do you think?

ewencp replied:
I like the idea of forcing connector developers to think about how upgrades will work when they need to change configurations. On the other hand, I'm not sure this will actually do anything -- people can easily just leave throw `UnsupportedOperationException` and ignore this whole process. I'm not convinced that adding these APIs actually helps since we aren't providing any patterns or code to help connector developers do this right, just adding API to tell them they should think about it.

That said, I don't feel strongly about this, so if people think it's worth it I can add those APIs.

gwenshap replied:
They can't throw unsupported upgrade - we will persist their version in the configuration and refuse to run if the versions don't match.

Yes, they can release new versions without bumping the version number, but thats going to some length to screw their users :)

I'll bring this up in the mailing list for discussion.

ewencp replied:
Where does the version for the update(from version) call come from? Does the user now have to include the version in the config or something? Or do you just record the version of the plugin you have when you first receive the connector config?

gwenshap commented on copycat-api/src/main/java/org/apache/kafka/copycat/util/ConnectorUtils.java:
```diff
@@ -0,0 +1,63 @@
+     * @param elements list of elements to partition
+     * @param numGroups the number of output groups to generate.
+     */
+    public static <T> List<List<T>> groupPartitions(List<T> elements, int numGroups) {
```
Nice :) I like how we support different optimizations without worrying about it too much.

gwenshap commented on copycat-api/src/test/java/org/apache/kafka/copycat/connector/ConnectorTest.java:
```diff
@@ -0,0 +1,76 @@
+
+import static org.junit.Assert.assertEquals;
+
+public class ConnectorTest {
```
We are not really testing the connector, right? We are testing the "reconfigure" part of the framework?

ewencp replied:
Yeah, I think this initially was planned to test more but ended up with very few tests since there isn't much to the base `Connector` class. Changed the name of the class to be clearer.

gwenshap commented on copycat-data/src/main/java/org/apache/kafka/copycat/data/ObjectProperties.java:
```diff
@@ -0,0 +1,85 @@
+
+
+
+package org.apache.kafka.copycat.data;
```
Since the data API is still under discussion and has followup JIRA, I'm ignoring the whole thing for now.
Makes sense? Is there any aspect you want me to review?

ewencp replied:
:+1: Ignoring it for now is the right choice.

gwenshap commented on copycat-file/src/main/java/org/apache/kafka/copycat/file/FileStreamSinkConnector.java:
```diff
@@ -0,0 +1,63 @@
+import java.util.Properties;
+
+/**
+ * Very simple connector that works with the console. This connector supports both source and
```
I like it. Short and sweet, like examples should be

ewencp replied:
One thing I realized is that now that we have Java 7 we could use the nio File classes for the Task implementations here. Think it's worth it or is it better to just keep this all as simple as possible?

ijuma replied:
I would use them if it makes the code simpler. Would in help in `FileStreamSourceTask` for example? I think the nio.2 APIs are generally saner (no booleans to indicate failure, for example), but on the con side people are less familiar with them.

gwenshap replied:
I'm good with either (i.e. I think writing to a file is simple enough in either API)

ewencp replied:
The source task is where it helps more -- see the hacky Thread.sleep in the current implementation. Regardless, this is easily handled in a later update.

gwenshap commented on copycat-file/src/main/java/org/apache/kafka/copycat/file/FileStreamSourceConnector.java:
```diff
@@ -0,0 +1,70 @@
+        filename = props.getProperty(FILE_CONFIG);
+        topic = props.getProperty(TOPIC_CONFIG);
+        if (topic == null || topic.isEmpty())
+            throw new CopycatException("ConsoleConnector configuration must include 'topic' setting");
```
SuperNit: The reference to console connector will be confusing when it comes from something called FileStreamSourceConnector...

gwenshap commented on copycat-file/src/main/java/org/apache/kafka/copycat/file/FileStreamSourceConnector.java:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    @Override
+    public List<Properties> getTaskConfigs(int maxTasks) {
```
maxTasks was in the connector configuration file, but in this case, any more than a single task will completely break the source. This should not be something that users can modify?

I keep bumping into the idea that there should be few logical "layers" of configurations. Will think about it a bit more.

ewencp replied:
I don't think we can avoid letting the user config this -- they need to be able to specify the maximum level of parallelism they want since each task gets its own thread. But `maxTasks` is meant to just be a suggestion, connectors can always ignore it if they simply cannot generate more tasks.

We could potentially require connectors to specify how many tasks they can generate, but I think this is only useful for a certain class of connectors that simply do not support parallelism at all (i.e. 1 partition in the source data).

gwenshap replied:
Makes sense. Basically allowing users to throttle connectors.

gwenshap commented on copycat-json/src/main/java/org/apache/kafka/copycat/json/JsonConverter.java:
```diff
@@ -0,0 +1,265 @@
+/**
+ * Implementation of Converter that uses JSON to store schemas and objects.
+ */
+public class JsonConverter implements Converter<JsonNode> {
```
This has the benefit of discouraging anyone who wants to implement their own convertor ;)

gwenshap commented on copycat-runtime/src/main/java/org/apache/kafka/copycat/cli/Copycat.java:
```diff
@@ -0,0 +1,106 @@
+ * limitations under the License.
+ **/
+
+package org.apache.kafka.copycat.cli;
```
I think we will want a CLI wrapper for the REST API at some point. Perhaps choose a name that will make it easier to understand the difference?

ewencp replied:
Not sure I understand the problem here. Couldn't the CLI for running standalone/distributed workers and for executing REST API calls live under this namespace? Or is it just that `cli` is not clear?

gwenshap replied:
yeah, same namespace is fine.

But right now "copycat --stuff" basically starts a single copycat flow and a single worker.
In the future "copycat --stuff" would do the same, but also submit jobs to the cluster, maybe report on cluster status, etc.

It made sense to me to separate the "run something here and now" and "control cluster" CLIs into two different commands.

gwenshap commented on copycat-runtime/src/main/java/org/apache/kafka/copycat/runtime/Coordinator.java:
```diff
@@ -0,0 +1,52 @@
+ * in workers. The assignment and tracking of those tasks in workers is also managed by the
+ * coordinator.
+ */
+public interface Coordinator {
```
This seems like a central component, but completely missing from KIP-26?

ewencp replied:
KIP-26 omitted a lot of implementation details. To be honest, I'm not certain how large this interface will get yet. I think it will have to contain most of the functionality that is exposed via the REST API since that implementation will essentially just be invoking methods on this class, but how large will that end up being.

Originally the thought for this class is that the DistributedCoordinator would be a significantly more complex implementation than the standalone version. However, if we are able to leverage the consumer coordinator code, then it may not even be that complex -- mostly just handles forwarding to the current leader (so, e.g., connector configs are only written by a single worker at any given time).

gwenshap replied:
I meant something a bit different - it is pretty unclear what the coordinator actually does (especially in stand-alone mode).

I want more detailed explanation, either here or in the KIP.

ewencp replied:
Ok, that makes sense. I've tried to clarify it a bit, see if the updated version makes more sense to you.

Another thing to think about -- coordinator is obviously a confusing name since there is already the consumer coordinator in Kafka. I would love to rename this class to something unique within Kafka, but am not coming up with anything better right now. Let me know if you have any ideas.

ewencp committed:
Address Gwen's review comments.

junrao commented on build.gradle:
```diff
@@ -204,20 +209,25 @@ for ( sv in ['2_10_5', '2_11_7'] ) {
 }
 
-tasks.create(name: "jarAll", dependsOn: ['jar_core_2_10_5', 'jar_core_2_11_7', 'clients:jar', 'examples:jar', 'contrib:hadoop-consumer:jar', 'contrib:hadoop-producer:jar', 'log4j-appender:jar', 'tools:jar']) {
+def copycatPkgs = ['copycat-data', 'copycat-api', 'copycat-runtime', 'copycat-json', 'copycat-file']
```
A quick comment on packaging. Could we put all copycat projects under a single directory (like what we did contrib/)? Also, is there a need to have different jars for api, data and runtime? It seems that they all need to be used together.

ewencp replied:
data + api can definitely be combined. I personally like using this separation to enforce clean layering, but there's not much risk of messing that up with those two.

The reason for splitting api and runtime is that connector developers should only need the api jar. Again, this is a way to really force proper layering, be clear about what we want to expose as public API to connector developers, etc.

gwenshap commented on copycat-runtime/src/main/java/org/apache/kafka/copycat/runtime/Worker.java:
```diff
@@ -0,0 +1,234 @@
+    private Converter<K> keyConverter;
+    private Converter<V> valueConverter;
+    private OffsetBackingStore offsetBackingStore;
+    private Serializer<K> offsetKeySerializer;
```
This will go on the mailing list, but I still think this is too much complexity too soon:
Offsets can be persisted with the data ser/de (if we think users care) or with our own format (if we think they don't). Separate pluggable serializers is complexity that doesn't buy much.

gwenshap commented on config/copycat-worker-local-persistent.properties:
```diff
@@ -0,0 +1,31 @@
+key.deserializer=org.apache.kafka.copycat.json.JsonDeserializer
+value.deserializer=org.apache.kafka.copycat.json.JsonDeserializer
+
+offset.storage.class=org.apache.kafka.copycat.storage.FileOffsetBackingStore
```
Maybe enum instead of class name... I don't if we want to encourage people to write their own offset stores.

ewencp replied:
Yeah, this was more to provide an easy way to switch between standalone mode (local, file-based offset storage) and distributed mode (stored in Kafka). But maybe we don't need a config for this since the implementation is fixed in each mode.

gwenshap replied:
Thank you!

gwenshap commented on copycat-runtime/src/main/java/org/apache/kafka/copycat/storage/MemoryOffsetBackingStore.java:
```diff
@@ -0,0 +1,114 @@
+ * behaves similarly to a real backing store, operations are executed asynchronously on a
+ * background thread.
+ */
+public class MemoryOffsetBackingStore implements OffsetBackingStore {
```
I assume this is just for testing?

ewencp committed:
Merge remote-tracking branch 'origin/trunk' into copycat

ewencp committed:
Reorganize Copycat code to put it all under one top-level directory.

ewencp committed:
Mark Copycat classes as unstable and reduce visibility of some classes where possible.

ewencp committed:
Merge remote-tracking branch 'origin/trunk' into copycat

ewencp committed:
Clarify and expand the explanation of the Copycat Coordinator interface.

ewencp commented:
@junrao I reorganized everything under a copycat directory, let me know if the new organization makes more sense. On merging data + api, given the discussion in https://issues.apache.org/jira/browse/KAFKA-2367, we may just drop the data package altogether. I personally prefer keeping the api in a separate jar from the runtime, which I think ensures very clear boundaries for connector developers.

@gwenshap Followed up on a bunch of your comments. I _think_ I've addressed all your comments that weren't going to be discussed more on the mailing list.

I also marked a bunch of public classes in copycat as unstable since we now have that annotation. I didn't mark _everything_ however. I just focused on the classes that are really public and "entry points" to the API. We can do this more comprehensively, but I wasn't sure it really makes sense (e.g. will marking `WorkerSinkTaskThread` as unstable really help anyone anyway?). Let me know if you think we should mark every class.

Finally, any thoughts on the exception hierarchy? Shall I get rid of the checked vs. runtime exception distinction? And shall I make CopycatException inherit from KafkaException? I'm thinking this may become important for embedded mode, where in some cases it would be nice to be able to catch any Copycat/Kstreams/Kafka client exception with a single catch block.

ewencp committed:
Make Copycat CLI speific to standalone mode, clean up some config and get rid of config storage in standalone mode.

gwenshap commented:
I agree there's no need to mark the stability of the internals. The whole point is to set expectations of external users.

Regarding exceptions:
- I don't have strong opinions on whether CopycatException should be KafkaException. I think my imagination failed on how Copycat will be used in embedded mode.
- Yeah, I think the checked vs. runtime is a bit unclear right now. Either document the specific distinctions with some guidelines on when to use each, or get rid of it for now.

And thank you for getting rid of the storage class config. I found that its easier to add configuration later than to remove it.

ijuma commented:
@ewencp, inheriting from KafkaException makes sense to me as I don't see any particular downsides. Do you?

Regarding checked versus unchecked exceptions, I think checked exceptions tend to be awkward as they don't compose (it's classified as a failed experiment by many). Unless the benefit is clear and we have clear guidelines on when we use checked versus when we use unchecked, I'd just drop the distinction.

ewencp commented:
Ok, I want to just drop the checked exceptions and I'll inherit from KafkaException since it unifies things and I think has little consequence. @gwenshap hold off on the merge until I have a chance to update these, please!

And thanks for the feedback everyone!

ewencp committed:
Rename Coordinator to Herder to avoid confusion with the consumer coordinator.

ewencp committed:
Simplify Copycat exceptions, make them a subclass of KafkaException.

becketqin commented:
Just went finish going through the code for first pass. Well structured code. A late comment, in which case would the runtime call Connector.reconfigure() instead of call ConnectorContext.requestTaskReconfiguration()? Also, would it worth having Initializable/Configurable/Startable/Shutdownable interface?

ewencp commented:
@becketqin `ConnectorContext.requestTaskReconfiguration` is for Connector implementations to call to indicate that some event happened that requires updating the tasks. For example, a JDBC connector would invoke it if it noticed a table was added since that would require work to be redistributed among tasks. `Connector.reconfigure` is called by the framework to update the configuration; for example, if a user submitted a modified connector configuration (say, added another connection URL and credentials to start copying another database, or added a table to the whitelist of tables), then the framework would invoke this to notify the connector of changes. The default implementation just stops and restarts the connector -- you only need to override `reconfigure` if stopping and restarting is too expensive. To be honest, I'm not sure this is valuable -- we may end up wanting to remove this just to keep things simpler, especially since we can add it with the default implementation in a later version if we want to.

On the interfaces, I considered it. The class hierarchy is already a bit more complex than I'd like and I didn't actually have a use for those interfaces. Since added `implements` to a class is backwards compatible, I figured it's better to only add them once we see a use for them. The only drawback to not adding them now is that if we end up with inconsistent method signatures we won't be able to add the interfaces in the future.
